File: python/valid_parentheses.py

#思路：用栈，遇到左括号入栈，遇到右括号检查栈顶是否匹配。最后栈为空则全部匹配。
from typing import List

def is_valid(s: str) -> bool:
    pairs = {')': '(', ']': '[', '}': '{'}
    stack: List[str] = []

    for ch in s:
        if ch in pairs:
            if not stack or stack[-1] != pairs[ch]:
                return False
            stack.pop()
        else:
            stack.append(ch)

    return not stack


# 测试
print(is_valid("()"))        # True
print(is_valid("()[]{}"))    # True
print(is_valid("(]"))        # False
print(is_valid("([])"))      # True
print(is_valid("["))         # False

File: TS/valid_parentheses.ts
function isValid(s: string): boolean {
    const pairs: Record<string, string> = {')': '(', ']': '[', '}': '{'};
    const stack: string[] = [];

    for (const ch of s) {
        if (ch in pairs) {
            if (!stack.length || stack[stack.length - 1] !== pairs[ch]) {
                return false;
            }
            stack.pop();
        } else {
            stack.push(ch);
        }
    }

    return stack.length === 0;
}

// 测试
console.log(isValid("()"));        // true
console.log(isValid("()[]{}"));    // true
console.log(isValid("(]"));        // false
console.log(isValid("([])"));      // true
console.log(isValid("["));         // false
